app/drivers/gpio/max7318/gpio_max7318.c

```diff
@@ -0,0 +1,310 @@
+        return -EWOULDBLOCK;
+    }
+
+    k_sem_take(&drv_data->lock, K_FOREVER);
+
+    uint16_t buf = drv_data->reg_cache.output;
+    buf = (buf & ~mask) | (mask & value);
+
+    int ret = write_registers(dev, REG_OUTPUT_PORTA, buf);
+    if (ret == 0) {
+        drv_data->reg_cache.output = buf;
+    }
+
+    k_sem_give(&drv_data->lock);
+    return ret;
+}
+
+static int max7318_port_set_bits_raw(const struct device *dev, uint32_t mask) {
+    return max7318_port_set_masked_raw(dev, mask, mask);
+}
+
+static int max7318_port_clear_bits_raw(const struct device *dev, uint32_t mask) {
+    return max7318_port_set_masked_raw(dev, mask, 0);
+}
+
+static int max7318_port_toggle_bits(const struct device *dev, uint32_t mask) {
+    struct max7318_drv_data *const drv_data = (struct max7318_drv_data *const)dev->data;
+
+    /* Can't do I2C bus operations from an ISR */
+    if (k_is_in_isr()) {
+        return -EWOULDBLOCK;
+    }
+
+    k_sem_take(&drv_data->lock, K_FOREVER);
+
+    uint16_t buf = drv_data->reg_cache.output;
+    buf ^= mask;
+
+    int ret = write_registers(dev, REG_OUTPUT_PORTA, buf);
+    if (ret == 0) {
+        drv_data->reg_cache.output = buf;
+    }
+
+    k_sem_give(&drv_data->lock);
+    return ret;
+}
+
+static int max7318_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+                                           enum gpio_int_mode mode, enum gpio_int_trig trig) {
+    return -ENOTSUP;
+}
+
+static const struct gpio_driver_api api_table = {
+    .pin_configure = max7318_config,
+    .port_get_raw = max7318_port_get_raw,
+    .port_set_masked_raw = max7318_port_set_masked_raw,
+    .port_set_bits_raw = max7318_port_set_bits_raw,
+    .port_clear_bits_raw = max7318_port_clear_bits_raw,
+    .port_toggle_bits = max7318_port_toggle_bits,
+    .pin_interrupt_configure = max7318_pin_interrupt_configure,
+};
+
+/**
+ * @brief Initialisation function of MAX7318
+ *
+ * @param dev Device struct
+ * @return 0 if successful, failed otherwise.
+ */
+static int max7318_init(const struct device *dev) {
+    const struct max7318_config *const config = dev->config;
+    struct max7318_drv_data *const drv_data = (struct max7318_drv_data *const)dev->data;
+
+    drv_data->i2c = device_get_binding((char *)config->i2c_dev_name);
+    if (!drv_data->i2c) {
+        LOG_DBG("Unable to get i2c device");
+        return -ENODEV;
+    }
+
+    if (!device_is_ready(drv_data->i2c)) {
+        LOG_WRN("i2c bus not ready!");
+        return -EINVAL;
+    }
+
+    LOG_INF("device initialised at 0x%x (i2c=%s)", config->device_addr, config->i2c_dev_name);
+
+    k_sem_init(&drv_data->lock, 1, 1);
+    return 0;
+}
+
+#define MAX7318_INIT(inst)                                                                         \
+    static struct max7318_config max7318_##inst##_config = {                                       \
+        .i2c_dev_name = DT_INST_BUS_LABEL(inst),                                                   \
```

Comment:
The common stuff should be initialized here, see https://github.com/zmkfirmware/zmk/blob/main/app/drivers/gpio/gpio_595.c#L200 for an example.

Also, Zephyr has newer APIs for the bus stuff, which isn't incorporated into our local mcp23017 driver yet, but should be used for newer drivers. The I2C idea can be seen at https://github.com/zmkfirmware/zmk/blob/main/app/drivers/gpio/gpio_595.c#L205 but there is an equivalent macro and API for SPI.

Reply:
I think you meant `s/spi/i2c/`. I hope I made the changes correctly (it compiles); will test with my board to make sure it works when I get the time (soon)